File: Problems/No.729_My_Calendar_I/my_calendar_I.cpp
class Event_Node {

public:
    int start_date, end_date;
    Event_Node *before, *after;
    Event_Node(int start_date, int end_date) {

        this->start_date = start_date;
        this->end_date = end_date;
        before = nullptr;
        after = nullptr;

    }

};

class MyCalendar {

private:
    Event_Node *root;

    bool check_valid(Event_Node *node, int start, int end) {

        if (node == nullptr) return true;
        else if (start >= node->end_date) return check_valid(node->after, start, end);
        else if (end <= node->start_date) return check_valid(node->before, start, end);
        else return false;

    }

    Event_Node *update_calendar(Event_Node *node, int start, int end) {

        if (node == nullptr) node = new Event_Node(start, end);
        else if (start >= node->end_date) node->after = update_calendar(node->after, start, end);
        else node->before = update_calendar(node->before, start, end);

        return node;

    }


public:

    MyCalendar() {root = nullptr;}
    
    bool book(int start, int end) {
        
        bool valid = check_valid(root, start, end);
        if (valid) root = update_calendar(root, start, end);
        return valid;

    }
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */

File: Problems/No.729_My_Calendar_I/my_calendar_I.py
class Event_Node:
    def __init__(self, start_date, end_date):
        self.start_date = start_date
        self.end_date = end_date
        self.before = None
        self.after = None
        

class MyCalendar(object):

    def __init__(self):
        self.root = None
    
    def check_valid(self, node, start, end):
        if not node:
            return True
        elif (start >= node.end_date):
            return self.check_valid(node.after, start, end)
        elif (end <= node.start_date):
            return self.check_valid(node.before, start, end)
        else:
            return False
    
    def update_calendar(self, node, start, end):
        if not node:
            node = Event_Node(start, end)
        elif (start >= node.end_date):
            node.after = self.update_calendar(node.after, start, end)
        else:
            node.before = self.update_calendar(node.before, start, end)
        return node

    def book(self, start, end):
        """
        :type start: int
        :type end: int
        :rtype: bool
        """

        valid = self.check_valid(self.root, start, end)
        if valid:
            self.root = self.update_calendar(self.root, start, end)
        return valid
        


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)
